federation/pkg/controller/service/servicecontroller.go

```diff
@@ -0,0 +1,956 @@
+	//}
+
+	return nil, notRetryable
+}
+
+// Returns whatever error occurred along with a boolean indicator of whether it
+// should be retried.
+func (s *ServiceController) ensureDNSRecordsExist(key string, fedService, appliedState *v1.Service) (error, bool) {
+	// Clone federation service, and create them in underlying k8s cluster
+	clone, err := conversion.NewCloner().DeepCopy(fedService)
+	if err != nil {
+		return nil, err
+	}
+	service, ok := clone.(*v1.Service)
+	if !ok {
+		return nil, fmt.Errorf("Unexpected service cast error : %v\n", service)
+	}
+	// to generate service name, we need a api.ObjectMeta instead of v1
+	meta := &api.ObjectMeta{}
+	meta.GenerateName = service.ObjectMeta.Name + "-"
+	service.Name = meta.Name
+	service.GenerateName = fedService.Name
+	service.ResourceVersion = ""
+
+	api.GenerateName(api.SimpleNameGenerator, meta)
+	// TODO: to generate labels based on hash
+
+	// handle available clusters one by one
+	for clusterName, client := range s.clusterClientMap {
+		go s.processServiceForCluster(key, clusterName, service, appliedState, client)
+	}
+}
+
+func retrieveExternalIPs(client clientset.Clientset, service v1.Service) {
+	// TODO: sleep and pull service info to check LB changes? or create k8s cluster service watch when create clusterClientMap
+	for i := 0; i < clientRetryCount; i++ {
+		time.Sleep(clientRetryInterval)
+		svc := getServiceByLabel(client, service)
+		if svc != nil {
+			return svc.Spec.ExternalIPs
+		}
+	}
+}
+
+func getServiceByLabel(client clientset.Clientset, service v1.Service) v1.Service{
+	// TODO: query service by service-spec-hash
+	svc, err := client.Core().Services(service.Namespace).List(&unversioned.LabelSelector{MatchLabels: map[string]string{serviceHashLabelName: "baz"}})
+	// One service in one cluster with same hash
+	if err == nil && svc != nil && len(svc.Items) == 0 {
+		return svc.Items[0]
+	} else {
+		return nil
+	}
+}
+
+func (s *ServiceController) ensureClusterServices(service *v1.Service, client clientset.Clientset, clusterName string) ([]string, error) {
+	var err error
+	for i := 0; i < clientRetryCount; i++ {
+		svc := getServiceByLabel(client, service)
+		if svc != nil {
+			// service exists
+			glog.V(4).Infof("service exists in cluster %v, trying to update", clusterName)
+			_, err = client.Core().Services(service.Namespace).UpdateStatus(service)
+			if err == nil {
+				// Update successfully, get LB IPs
+				return retrieveExternalIPs(client, service)
+			}
+		}
+
+		if errors.IsNotFound(err) {
+			// Create service if is not found
+			glog.Infof("Service '%s/%s' does not exist in %s, trying to create new",
+				service.Namespace, service.Name, clusterName)
+			_, err = client.Core().Services(service.Namespace).Create(service)
+			if err == nil {
+				return retrieveExternalIPs(client, service)
+			}
+		}
+
+		if errors.IsConflict(err) {
+			glog.V(4).Infof("Not persisting update to service '%s/%s' that has been changed since we received it: %v",
```

Comment:
I think that we should perhaps retry here (i.e. do a get and update), rather than fail.
Of would you prefer to have the next reconciliation loop do that retry?